lib/filewriter.js

```diff
@@ -0,0 +1,182 @@
+        else return this._reopenFd(callback);
+    };
+
+    this._getLockedFd = function(callback) {
+        var self = this;
+        this._getFd(function(err, fd) {
+            if (err) return callback(err);
+            flock(fd, "ex", function(err) {
+                if (err) return callback(err);
+                if (Date.now() < self.reopenTime) {
+                    // grab and lock the fd and only then test reopenTime, since
+                    // the expectation is that most writes will be bunched
+                    // NOTE: this is also needed for correct renameFile
+                    return callback(null, fd);
+                }
+                else {
+                    flockSync(fd, "un");
+                    self._reopenFd(function(err, fd) {
+                        if (err) return callback(err);
+                        self._getLockedFd(callback);
+                    });
+                }
+            });
+        })
+    };
+
+    // atomically append the string to the file
+    this.write = function(str, cb) {
+        var self = this;
+        this._getLockedFd(function(err, fd) {
+            if (err) return cb(err);
+            // not much benefit to reusing a buffer for the writes
+            var buf = new Buffer(str);
+            fs.write(fd, buf, 0, buf.length, null, function(err, nb) {
+                flockSync(fd, "un");
+                cb(err, nb);
+            });
+        });
+    };
+
+    // close the file
+    this.close = function( ) {
+        try { if (this.fd) fs.closeSync(this.fd); }
+        catch (err) { }
+        this.fd = undefined;
+    };
+
+    // expose class method on instance
+    this.mutexTimeout = FileWriter.mutexTimeout;
+    this.renameFile = FileWriter.renameFile;
+}
+
+
+// rename the file and wait for write activity to cease
+FileWriter.mutexTimeout = 5000;
+FileWriter.renameFile = function renameFile( oldName, newName, waitMs, cb ) {
+    if (!cb && typeof waitMs === 'function') {
+        cb = waitMs
+        waitMs = 50
+    }
+    // Wait for all pending writes to finish.  After waitMs no writer will reuse the old fd,
+    // and once we`ve paused and have the mutex we know that no more writes will occur.
+    var fd = null, mutexTimeout = (this && this.mutexTimeout) ? this.mutexTimeout : FileWriter.mutexTimeout;
+    function waitForWritesToFinish( ) {
+        var guard, error = null
+        function finish(err) {
+            if (fd !== null) fs.close(fd, function(){})
+            fd = null
+            if (guard && global.clearTimeout) clearTimeout(guard)
+            guard = null
+            return error ? cb(error) : cb(err)
+        }
+        try {
+            guard = setTimeout(function() {
+                if (fd !== null) fs.close(fd, function(){})
+                error = new Error("timed out waiting for last write")
+                return finish(error)
+            }, mutexTimeout)
+            if (guard.unref) guard.unref()
```

Comment:
hmm... no, it's not necessary.  Not because of efficiency concerns (detaching a timer
every now and then will not be measurable), but because the point to unref-ing the guard
is to not hang the program if it wants to exit.  But the guard only exists until the flock finishes,
and if the flock() doesn't return the program will hang on that anyway, ie the unref is pointless.